packages/cicero-core/src/template.js

```diff
@@ -172,15 +179,61 @@ class Template {
         return hasher.digest('hex');
     }
 
+    /**
+     * signs a string made up of template hash and time stamp using private key derived
+     * from the keystore
+     * @param {string} [path] - path of the keystore to be used
+     * @param {string} [password] - password for the keystore file
+     * @param {string} [timeStamp] - timeStamp of the moment of signature is done
+     * @return {object} - object containing template hash, timestamp, signatory's certificate, signature
+     * @private
+     */
+    signTemplate(path, password, timeStamp) {
+        const templateHash = this.getHash();
+        const p12Ffile = fs.readFileSync(path, { encoding: 'base64' });
+        // decode p12 from base64
+        const p12Der = forge.util.decode64(p12Ffile);
+        // get p12 as ASN.1 object
+        const p12Asn1 = forge.asn1.fromDer(p12Der);
+        // decrypt p12 using the password 'password'
+        const p12 = forge.pkcs12.pkcs12FromAsn1(p12Asn1, false, password);
+        //X509 cert forge type
+        const certificateForge = p12.safeContents[0].safeBags[0].cert;
+        //Private Key forge type
+        const privateKeyForge = p12.safeContents[1].safeBags[0].key;
+        //convert cert and private key from forge to PEM
+        const certificatePem = forge.pki.certificateToPem(certificateForge);
+        const privateKeyPem = forge.pki.privateKeyToPem(privateKeyForge);
+        //convert private key in pem to private key type in node
+        const privateKey = crypto.createPrivateKey(privateKeyPem);
+        const sign = crypto.createSign('SHA256');
+        sign.write(templateHash + timeStamp);
+        sign.end();
+        const signature = sign.sign(privateKey, 'hex');
+        const signatureObject = {
+            templateHash: templateHash,
+            timeStamp: timeStamp,
+            signatoryCert: certificatePem,
+            signature: signature
+        };
+        return signatureObject;
+    }
+
     /**
      * Persists this template to a Cicero Template Archive (cta) file.
      * @param {string} [language] - target language for the archive (should be 'ergo')
+     * @param {Object} [keyStore] - path for the keystore and password of the keystore
      * @param {Object} [options] - JSZip options
-     * @param {Buffer} logo - Bytes data of the PNG file
      * @return {Promise<Buffer>} the zlib buffer
      */
-    async toArchive(language, options) {
-        return TemplateSaver.toArchive(this, language, options);
+    async toArchive(language, keyStore, options) {
+        if(this.authorSignature === null){
```

Comment:
Please write `if (this.authorSignature === null) {` and likewise below `} else {` for readability and consistency with the rest of the code.

Reply:
Possibly `if (!this.authorSignature)` might be best in this case instead of `this.authorSignature === null`.

Reply:
Since the template shouldn't be signed if the `keyStore` value is empty, I think the following structure would be better:
```
// do not return signature, set authorSignature instead
signTemplate(path, password, timeStamp) {
   ... // existing code
   this.authorSignature = signatureObject;
}

async toArchive(language, options) { // leave the signature part in the options
  if (options.keyStore) {
    // This is executed when signature is required
    if (!this.authorSignature) {
      const timeStamp = Date.now();            
      this.signTemplate(options.keyStore.path, options.keyStore.password, timeStamp);
    } else {
      // This is executed if signature is required but already present
      // Q: don't we want to allow to sign again?
      throw new Error('Template is already signed by the author.');
    }
  }
  // This is always executed
  return TemplateSaver.toArchive(this, language, options);
}
```

Reply:
> Q: don't we want to allow to sign again?

I wanted to allow only the original author to sign the template. I don't know if it's needed or not. Might have to modify the code such that after some change the author should be able to do the signature agiain. If you think we don't need that I'll remove that.

Reply:
@jeromesimeon what do you suggest for this?